modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+  int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+  return true; // to be done
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+// this is a simple algorithm simply based on the added angle = 360 degrees
+// works fine for planar cutting surface
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0);
+  std::vector<unsigned int> node_id(_cut_elem_nnode);
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+
+    // To match the type definition and data structure in Xfem::angle_rad_3d
+    // C array is used here instead of std::array or std::vector
+    Real vertices[_cut_elem_nnode][LIBMESH_DIM];
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (std::abs(angle[i] - libMesh::pi * 2) > libMesh::pi * 2 * 0.001)
+    {
+      std::shared_ptr<CutPoint> ptr(new CutPoint);
```

Comment:
`new CutPoint` -> `std::make_shared`

Reply:
After changing "std::shared_ptr<CutPoint> ptr(new CutPoint)" to "std::shared_ptr<CutPoint> ptr = std::make_shared<CutPoint>", I got the following errors.  What would be the cleanest way to fix this? 

/Users/liu/projects/moose/modules/xfem/src/userobjects/MeshCut3DUserObject.C:419:33: error: no viable conversion from '<overloaded function type>' to
      'std::shared_ptr<CutPoint>'
      std::shared_ptr<CutPoint> ptr = std::make_shared<CutPoint>;
                                ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3879:23: note: candidate constructor not viable: no overload of 'make_shared' matching 'nullptr_t' for
      1st argument
    _LIBCPP_CONSTEXPR shared_ptr(nullptr_t) _NOEXCEPT;
                      ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3893:5: note: candidate constructor not viable: no overload of 'make_shared' matching
      'const std::__1::shared_ptr<CutPoint> &' for 1st argument
    shared_ptr(const shared_ptr& __r) _NOEXCEPT;
    ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3901:5: note: candidate constructor not viable: no overload of 'make_shared' matching
      'std::__1::shared_ptr<CutPoint> &&' for 1st argument
    shared_ptr(shared_ptr&& __r) _NOEXCEPT;
    ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3896:9: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-0>' against
      'std::__1::shared_ptr<CutPoint> ()'
        shared_ptr(const shared_ptr<_Yp>& __r,
        ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3902:52: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-0>' against
      'std::__1::shared_ptr<CutPoint> ()'
    template<class _Yp> _LIBCPP_INLINE_VISIBILITY  shared_ptr(shared_ptr<_Yp>&& __r,
                                                   ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3910:9: note: candidate template ignored: could not match 'auto_ptr<type-parameter-0-0>' against
      'std::__1::shared_ptr<CutPoint> ()'
        shared_ptr(auto_ptr<_Yp>&& __r,
        ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3919:9: note: candidate template ignored: could not match 'unique_ptr<type-parameter-0-0,
      type-parameter-0-1>' against 'std::__1::shared_ptr<CutPoint> ()'
        shared_ptr(unique_ptr<_Yp, _Dp>&&,
        ^
/opt/moose/llvm-3.9.0/bin/../include/c++/v1/memory:3928:9: note: candidate template ignored: could not match 'unique_ptr<type-parameter-0-0,
      type-parameter-0-1>' against 'std::__1::shared_ptr<CutPoint> ()'
        shared_ptr(unique_ptr<_Yp, _Dp>&&,
        ^
1 error generated.
make: *** [/Users/liu/projects/moose/modules/xfem/src/userobjects/MeshCut3DUserObject.x86_64-apple-darwin16.7.0.opt.lo] Error 1